bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/main/java/com/google/cloud/bigtable/mapreduce/validation/SyncTableValidationOnlyJob.java

```diff
@@ -0,0 +1,1102 @@
+                + sourceConf.get(ClusterConnection.HBASE_CLIENT_CONNECTION_IMPL)
+                + ", "
+                + sourceConf.get(BigtableOptionsFactory.BIGTABLE_EMULATOR_HOST_KEY)
+                + ". Source config HBase zk: "
+                + sourceConf.get(SOURCE_ZK_CLUSTER_CONF_KEY)
+                + ", and Bigtable project: "
+                + sourceConf.get(SOURCE_BT_PROJECTID_CONF_KEY)
+                + ", instance: "
+                + sourceConf.get(SOURCE_BT_INSTANCE_CONF_KEY)
+                + ", table: "
+                + sourceConf.get(SOURCE_TABLE_CONF_KEY)
+                + " with error: "
+                + e.getMessage()
+                + ". Processing continuing without connecting to source",
+            e);
+      }
+
+      sourceTableHash = HashTable.TableHash.read(conf, sourceHashDir);
+      LOG.info("Read source hash manifest: " + sourceTableHash);
+      LOG.info(
+          "Read "
+              + BigtableTableHashAccessor.getPartitions(sourceTableHash).size()
+              + " partition keys");
+
+      TableSplit split = (TableSplit) context.getInputSplit();
+      ImmutableBytesWritable splitStartKey = new ImmutableBytesWritable(split.getStartRow());
+
+      sourceHashReader = sourceTableHash.newReader(conf, splitStartKey);
+      findNextKeyHashPair();
+
+      // create a hasher, but don't start it right away
+      // instead, find the first hash batch at or after the start row
+      // and skip any rows that come before.  they will be caught by the previous task
+      targetHasher = new BigtableResultHasher();
+    }
+
+    /**
+     * create connection to data store. if zkClusterConfKey is null or empty string, the connection
+     * will default to properties set in conf for establishing connection.
+     *
+     * @param conf configuration for connection
+     * @param zkClusterConfKey zookeeper configuration key for source or target
+     * @return Connection to cluster
+     * @throws IOException on connection
+     */
+    private static Connection openConnection(Configuration conf, String zkClusterConfKey)
+        throws IOException {
+      String zkCluster = conf.get(zkClusterConfKey);
+      Configuration clusterConf = HBaseConfiguration.createClusterConf(conf, zkCluster);
+      return ConnectionFactory.createConnection(clusterConf);
+    }
+
+    private static Table openTable(
+        Connection connection, Configuration conf, String tableNameConfKey) throws IOException {
+      return connection.getTable(TableName.valueOf(conf.get(tableNameConfKey)));
+    }
+
+    /**
+     * Attempt to read the next source key/hash pair. If there are no more, set nextSourceKey to
+     * null
+     */
+    private void findNextKeyHashPair() throws IOException {
+      boolean hasNext = sourceHashReader.next();
+      if (hasNext) {
+        nextSourceKey = sourceHashReader.getCurrentKey();
+      } else {
+        // no more keys - last hash goes to the end
+        nextSourceKey = null;
```

Comment:
When I was reading the hbase code, it was my impression that both the ends were represented by EMPTY_BYTE_ARRAY, will making it null make it incompatible with vanilla hashtable ?